### ISSUE TITLE
Rule proposal: no-code-in-comments

### ISSUE THREAD

**Issue by Skalman** [eslint/eslint#2735]
While developing, I frequently try different solutions, and end up with commented-out code, which I occasionally even commit. I'd like for ESLint to warn me about this.

## How to detect code
A line of code ends with one of the following characters: `{};` .
* Advantages: Simple.
* Disadvantages: Would miss something like `//console.log('test')`, and while testing code ignoring the semicolon is probably common.

An alternative would be to somehow detect valid JS, but that seems unnecessarily complex.

## Reporting
Adjacent comments should generate only a single error. You can view the following as a single comment block.
```js
// console.log(1);
// console.log(2);
// console.log(3);
```

## Options
Inline documentation can sometimes contain code, so we need a way to allow some comments, perhaps in comments starting with `/**`.

Suggested default:
```js
[2, {isCodeRegExp: /[\{\};](\n|$)/, allowRegExp: /^\*/}]
```
(The options will be transformed into regexps when strings are given.)

**Comment by lo1tuma** [comment]
Duplicate of #604.

**Comment by nzakas** [comment]
We closed #604 as being too difficult to make work correctly. Your proposal has some significant blind spots (not working unless you use semicolons is a big blind spot).

Unless you have an idea for how to make this more reliable, I don't see this moving forward.

**Comment by Skalman** [comment]
The way I see it, it's better to warn about most of the commented code, even if we can't catch all. The important thing is that we don't catch cases that *aren't* code. Here are some other patterns that are common in JS, but pretty much non-existent in English, and I believe other languages too:

```js
// function calls: a(), b2(), c.d(e), $(sel)
/[a-zA-Z0-9_\$]\(/
   // Note: Maybe we don't want to catch this. A reasonable comment:

   // a() with no parameters will always return true.


// operators: a && b || c  (could probably add some others too)
/&&|\|\|/


// conditional constructs: if (true) return
/(^|\n)[ \t]*(if|for|while) \(/
   // Can it be improved? Currently also catches:

   // Long line long line long line. Do this
   // if (and only if) some condition is true.


// comment contains a comment: // // I'm commented twice (well, technically three times)
/[^:]\/\//
   // Allow links: http://...
```

Here is some normal, commented-out code that the above regexes suggestions *won't* catch, but I don't see you could catch them without lot's of false positives:
```js
var a = [
  1,
  //2,
  3
];

// location.href = ''
```

**Comment by lo1tuma** [comment]
What about an example code in comment, this should not be a warning:

```js
// Example usage: foo.bar("baz");
```

**Comment by nzakas** [comment]
@Skalman I'm sorry, I'm very uncomfortable with regex based matching for this. Too many possibilities for problems.

Of course, you can always write a custom rule that does this for your own code, I just don't see this going into our core rules.

**Comment by tmsdnl** [comment]
Came across a custom rule by @bahmutov: 
https://github.com/bahmutov/eslint-rules/blob/master/no-commented-out-code.js

Any thoughs?

**Comment by ilyavolodin** [comment]
I saw that plugin a while back, and it basically runs every comment through espree. If it gets valid AST - it's commented out code. Problem is, you might have commented out just a few statements, for example, just a `case` block inside a `switch` which will not parse properly, but it is still a commented out code.
